Sources/Flows/OAuth2PasswordGrant.swift

```diff
@@ -37,40 +53,114 @@ open class OAuth2PasswordGrant: OAuth2 {
-		username = settings["username"] as? String ?? ""
-		password = settings["password"] as? String ?? ""
+		username = settings["username"] as? String
+		password = settings["password"] as? String
 		super.init(settings: settings)
 	}
 	
-	override open func doAuthorize(params: [String : String]? = nil) {
-		self.obtainAccessToken(params: params) { params, error in
-			if let error = error {
-				self.didFail(with: error)
+	/**
+	Performs the accessTokenRequest if credentials are already provided, or ask for them with a native controller.
+	*/
+	override open func doAuthorize(params: OAuth2StringDict? = nil) throws {
+		
+		authorizationResponse = nil
+		
+		if username?.isEmpty ?? true || password?.isEmpty ?? true {
+			try askForCredentials()
+		} else {
+			obtainAccessToken(params: params) { params, error in
+				if let error = error {
+					self.didFail(with: error)
+				} else {
+					self.didAuthorize(withParameters: params ?? OAuth2JSON())
+				}
 			}
-			else {
-				self.didAuthorize(withParameters: params ?? OAuth2JSON())
+		}
+	}
+	
+	/**
+	Present the delegate's loginController to the user.
+	*/
+	private func askForCredentials(params: OAuth2StringDict? = nil) throws {
+		logger?.debug("OAuth2", msg: "Presenting the login controller")
+		
+		guard let delegate = delegate else {
+			throw OAuth2Error.noDelegate
+		}
+		
+		try loginPresenter.present(loginController: delegate.loginController(oauth2: self),
+								   fromContext: authConfig.authorizeContext,
+								   animated: true)
+		additionalParams = params
+	}
+	
+	/**
+		Submits loginController's provided credentials to the OAuth server.
+		The completionHandler is called once the server responded with the appropriate error or `nil` is the user is
+		now authorized.
+		This doesn't automatically call `endAuthorization` once the user is authorized, allowing the login controller to
+		perform any kind of confirmation before its dismissal.
+	*/
+	public func tryCredentials(username: String,
+							   password: String,
+							   completionHandler: @escaping (OAuth2JSON?, OAuth2Error?) -> Void) {
+		
+		self.username = username
+		self.password = password
+		
+		//Perform the request
+		obtainAccessToken(params: additionalParams, callback: { params, error in
+			//Reset credentials
+			if error != nil {
+				self.username = nil
+				self.password = nil
 			}
+			
+			completionHandler(params, error)
+		})
+	}
+	
+	/**
+	Ends the authorization process whether the user had been successfully authorized or not.
+	*/
+	public func endAuthorization(animated: Bool = true) {
+		//Some clean up
+		additionalParams = nil
+		
+		logger?.debug("OAuth2", msg: "Dismissing the login controller")
+		loginPresenter.dismissLoginController(animated: animated)
+		
+		//Call the right authorization callback according to the last server response
+		if let response = authorizationResponse {
+			self.didAuthorize(withParameters: response)
+		} else {
+			self.didFail(with: nil)
```

Comment:
I think we actually need an hybrid solution, mirroring how we actually log in using a webView:
the webview handle errors by itself, but is autoDismissed after a success.

How about having 2 callback: 
- the global `authorize` callback, notifying when the whole authorization process succeeded or failed. 
- a partial `tryCredentials`error callback, notifying only about accessToken request errors.

If the accessToken request returns with an error, the login controller isn't automatically dismissed and the partial error callback is called, allowing to show the error as the webview would do it.
If the request returns with no error, the login controller is automatically dismissed and the global callback is called.
This way, the callbacks now have different purposes and are no longer redundant.

I strongly don't recommend using only 1 callback because the object calling `authorize` would handle the login's controller, while the login controller is more capable of handling it by itself.